google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java

```diff
@@ -0,0 +1,568 @@
+                    update(LocalFirestoreHelper.SINGLE_FIELD_PROTO, "coll/doc2"),
+                    create(LocalFirestoreHelper.SINGLE_FIELD_PROTO, "coll/doc3")),
+                mergeResponses(successResponse(0), successResponse(1), successResponse(2)));
+          }
+        };
+    requestResponseMap.initializeStub(batchWriteCapture, firestoreMock);
+
+    bulkWriter.setMaxBatchSize(3);
+    ApiFuture<WriteResult> result1 = bulkWriter.set(doc1, LocalFirestoreHelper.SINGLE_FIELD_MAP);
+    ApiFuture<WriteResult> result2 = bulkWriter.update(doc2, LocalFirestoreHelper.SINGLE_FIELD_MAP);
+    ApiFuture<WriteResult> result3 =
+        bulkWriter.create(
+            firestoreMock.document("coll/doc3"), LocalFirestoreHelper.SINGLE_FIELD_MAP);
+
+    // The 4th write should not be sent because it should be in a new batch. However, this write
+    // should trigger the first batch to be sent.
+    bulkWriter.delete(firestoreMock.document("coll/doc4"));
+
+    List<BatchWriteRequest> requests = batchWriteCapture.getAllValues();
+    assertEquals(requestResponseMap.size(), requests.size());
+
+    verifyRequests(requests, requestResponseMap);
+    assertEquals(Timestamp.ofTimeSecondsAndNanos(0, 0), result1.get().getUpdateTime());
+    assertEquals(Timestamp.ofTimeSecondsAndNanos(1, 0), result2.get().getUpdateTime());
+    assertEquals(Timestamp.ofTimeSecondsAndNanos(2, 0), result3.get().getUpdateTime());
+  }
+
+  @Test
+  public void doesNotSendBatchesIfSameDocIsInFlight() throws Exception {
+    final RequestResponseMap requestResponseMap =
+        new RequestResponseMap() {
+          {
+            put(
+                batchWrite(
+                    set(LocalFirestoreHelper.SINGLE_FIELD_PROTO, "coll/doc1"),
+                    set(LocalFirestoreHelper.SINGLE_FIELD_PROTO, "coll/doc2")),
+                mergeResponses(successResponse(1), successResponse(2)));
+            put(
+                batchWrite(set(LocalFirestoreHelper.SINGLE_FIELD_PROTO, "coll/doc1")),
+                successResponse(3));
+          }
+        };
+    requestResponseMap.initializeStub(batchWriteCapture, firestoreMock, true);
+    bulkWriter.set(doc1, LocalFirestoreHelper.SINGLE_FIELD_MAP);
+    bulkWriter.set(doc2, LocalFirestoreHelper.SINGLE_FIELD_MAP);
+
+    // Schedule flush on separate thread to avoid blocking main thread while waiting for
+    // activeRequestComplete.
+    ScheduledFuture<ApiFuture<Void>> flush1 =
+        Executors.newSingleThreadScheduledExecutor()
+            .schedule(
+                new Callable<ApiFuture<Void>>() {
+                  public ApiFuture<Void> call() {
+                    return bulkWriter.flush();
+                  }
+                },
+                0,
+                TimeUnit.MILLISECONDS);
+
+    // Wait for flush() to perform logic and reach the stubbed response. This simulates a first
+    // batch that has been sent with its response still pending.
+    requestResponseMap.awaitLatch();
```

Comment:
Please use a more descriptive name for this API that makes it obvious what this latch does. Would `awaitRequest`/`awaitNextRequest` work?